### PR TITLE
[Tune] Add example and tutorial for DCGAN

### PR DESCRIPTION
**Pull request by hhbyyh**
Add the tutorial and examples
## Related issue number
https://github.com/ray-project/ray/issues/5660

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: doc/source/index.rst [hunk 1]
```diff
@@ -261,6 +261,7 @@ Getting Involved
    tune-schedulers.rst
    tune-searchalg.rst
    tune-package-ref.rst
+   tune-advance-tutorial.rst
```
Comment (richardliaw) [inline doc/source/index.rst]: let's put this right after the walkthrough

File: doc/source/tune-advance-tutorial.rst [hunk 2]
```diff
@@ -0,0 +1,251 @@
+Tune Advance tutorial
```
Comment (richardliaw) [inline doc/source/tune-advance-tutorial.rst]: Tune Advanced Tutorials

File: doc/source/tune-advance-tutorial.rst [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+=====================
+
+In this page, we will explore some advanced functions in Tune with more examples.
+
+Current available tutorials:
```
Comment (richardliaw) [inline doc/source/tune-advance-tutorial.rst]: On this page:

File: doc/source/tune-advance-tutorial.rst [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+    import ray
+    from ray import tune
+    from ray.tune import Trainable
+
+    class Guesser(Trainable):
+        def _setup(self, config):
+            self.config = config
+            self.password = 1024
+            pass
```
Comment (richardliaw) [inline doc/source/tune-advance-tutorial.rst]: no need for pass

File: doc/source/tune-advance-tutorial.rst [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        config={
+            "guess": tune.randint(1, 10000)
+        })
+
+    print('best config: ', analysis.get_best_config(metric="diff", mode="min"))
+
+The program randomly picks 10 number from [1, 10000) and finds which is closer to the password.
+As a subclass of ``ray.tune.Trainable``, Tune will convert ``Example`` into a Ray actor, which
+runs on a separate process on a worker. ``_setup`` function is invoked once for each Actor for custom
+initialization. ``_train`` execute one logical iteration of training in the tuning process,
```
Comment (richardliaw) [inline doc/source/tune-advance-tutorial.rst]:
```suggestion
initialization. 

``_train`` execute one logical iteration of training in the tuning process,
```

File: doc/source/tune-advance-tutorial.rst [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+which may include several iterations of actual training (see the next example). As a rule of
+thumb, the execution time of one train call should be large enough to avoid overheads
+(i.e. more than a few seconds), but short enough to report progress periodically
+(i.e. at most a few minutes).
+
+We only implemented ``_setup`` and ``_train``methods for simplification, usually it's also required
```
Comment (richardliaw) [inline doc/source/tune-advance-tutorial.rst]:
```suggestion
We only implemented ``_setup`` and ``_train`` methods for simplification, usually it's also required
```

File: doc/source/tune-advance-tutorial.rst [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+to implement ``_save``, and ``_restore`` for checkpoint and fault tolerance.
+
+Next we trains a Pytorch convolution model with Trainable and PBT.
```
Comment (richardliaw) [inline doc/source/tune-advance-tutorial.rst]:
```suggestion
Next, we train a Pytorch convolution model with Trainable and PBT.
```

File: doc/source/tune-advance-tutorial.rst [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+
+Trainable with Population Based Training (PBT)
+----------------------------------------------
+
+Tune includes a distributed implementation of `Population Based Training (PBT) <https://deepmind.com/blog/population-based-training-neural-networks>`__ as
+a scheduler `PopulationBasedTraining <tune-schedulers.html#Population Based Training (PBT)>`__ .
```
Comment (richardliaw) [inline doc/source/tune-advance-tutorial.rst]: Let's add a picture for PBT here.

File: doc/source/tune-advance-tutorial.rst [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+PBT starts by training many neural networks in parallel with random hyperparameters. But instead of the
+networks training independently, it uses information from the rest of the population to refine the
+hyperparameters and direct computational resources to models which show promise. This takes its inspiration
```
Comment (richardliaw) [inline doc/source/tune-advance-tutorial.rst]: Can we keep *paragraphs* to no more than 2 sentences?

It's ok to have multiple paragraphs here.

File: doc/source/tune-advance-tutorial.rst [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+
+.. literalinclude:: ../../python/ray/tune/examples/pbt_convnet_example.py
+   :language: python
+   :start-after: __trainable_begin__
+   :end-before: __trainable_end__
+
+The example reuses some of the functions in ray/tune/examples/mnist_pytorch.py, and is also a good
+demo for how to decouple the tuning logic and original training code.
+
+Here we also overrides ``reset_config``. This method is optional, but can be implemented to speed
```
Comment (richardliaw) [inline doc/source/tune-advance-tutorial.rst]:
```suggestion
Here, we also override ``reset_config``. This method is optional but can be implemented to speed
```

File: doc/source/tune-advance-tutorial.rst [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+up algorithms such as PBT, and to allow performance optimizations such as running experiments
+with reuse_actors=True.
```
Comment (richardliaw) [inline doc/source/tune-advance-tutorial.rst]:
```suggestion
with ``reuse_actors=True``.
```

File: doc/source/tune-advance-tutorial.rst [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+
+Then we define a PBT scheduler
+
+.. literalinclude:: ../../python/ray/tune/examples/pbt_convnet_example.py
+   :language: python
+   :start-after: __pbt_begin__
+   :end-before: __pbt_end__
+
+Some of the most important parameters are:
+``hyperparam_mutations`` and ``custom_explore_fn`` are used to mutate the hyperparameters.
```
Comment (richardliaw) [inline doc/source/tune-advance-tutorial.rst]: maybe a bullet list here?

File: python/ray/tune/examples/dcgan_mnist_pbt/dcgan_mnist_pbt.py [hunk 13]
```diff
@@ -0,0 +1,364 @@
...
+                                     betas=(beta1, 0.999))
+        self.dataloader = get_data_loader()
+
+    def _train(self):
+        lossG, lossD, is_score = train(self.netD, self.netG, self.optimizerG,
+                                       self.optimizerD, self.criterion,
+                                       self.dataloader, self._iteration)
+        return {"lossg": lossG, "lossd": lossD, "is_score": is_score}
+
+    def _save(self, checkpoint_dir):
```
Comment (richardliaw) [inline python/ray/tune/examples/dcgan_mnist_pbt/dcgan_mnist_pbt.py]: Ah, one thing that we did not do correctly in other examples is that we need to save the state_dict of the optimizer.

``self.optimizerG.state_dict()`` and ``self.optimizerD.state_dict()``

**Review by richardliaw — COMMENTED** [review]
This is awesome! Thank you very much for this contribution. 

I've left a few comments, but let's try to merge this ASAP and then we can iterate on it.

File: python/ray/tune/examples/dcgan_mnist_pbt/dcgan_mnist_pbt.py [hunk 14]
```diff
@@ -0,0 +1,364 @@
...
+from torch.nn import functional as F
+import torch.utils.data
+from scipy.stats import entropy
+
+# Set random seed for reproducibility
+manualSeed = 999
+# manualSeed = random.randint(1, 10000) # use if you want new results
+print("Random Seed: ", manualSeed)
+random.seed(manualSeed)
+torch.manual_seed(manualSeed)
```
Comment (richardliaw) [inline python/ray/tune/examples/dcgan_mnist_pbt/dcgan_mnist_pbt.py]: btw, does this actually work? I think the seed needs to be set per trainable (i.e., in _setup)

File: python/ray/tune/examples/dcgan_mnist_pbt/dcgan_mnist_pbt.py [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+        real_batch[0].to(device)[:64], padding=2, normalize=True).cpu(), (1, 2, 0)))
+
+    if not args.smoke_test:
+        plt.show()
+
+    # load the pretrained mnist classification model for inception_score
+    mnist_cnn = Net()
+    mnist_cnn.load_state_dict(torch.load("./mnist_cnn.pt"))
+    mnist_cnn.eval()
+    mnist_model_ref = ray.put(mnist_cnn)
```
Comment (richardliaw) [inline python/ray/tune/examples/dcgan_mnist_pbt/dcgan_mnist_pbt.py]: Instead of this, I think you can just specify a `restore=` parameter in `tune.run`

Comment (richardliaw) [reply python/ray/tune/examples/dcgan_mnist_pbt/dcgan_mnist_pbt.py]: oops nvm, misread

File: doc/source/index.rst (hunk 1 above)
Comment (hhbyyh) [reply doc/source/index.rst]: sure. Also fix the typo --> 'advanced'

File: python/ray/tune/examples/dcgan_mnist_pbt/dcgan_mnist_pbt.py (hunk 14 above)
Comment (hhbyyh) [reply python/ray/tune/examples/dcgan_mnist_pbt/dcgan_mnist_pbt.py]: remove it for now.

File: python/ray/tune/examples/dcgan_mnist_pbt/dcgan_mnist_pbt.py (hunk 13 above)
Comment (hhbyyh) [reply python/ray/tune/examples/dcgan_mnist_pbt/dcgan_mnist_pbt.py]: Sure. Updated.

**Comment by hhbyyh** [comment]
Not sure why but I guess rebase and the commit from Github webpage are not working well together. I'll try to close this one and open a new one...

**Comment by hhbyyh** [comment]
new PR: https://github.com/ray-project/ray/pull/6400
Sorry for the inconvenience.

File: doc/source/tune-advance-tutorial.rst (hunk 8 above)
Comment (hhbyyh) [reply doc/source/tune-advance-tutorial.rst]: done

File: doc/source/tune-advance-tutorial.rst (hunk 9 above)
Comment (hhbyyh) [reply doc/source/tune-advance-tutorial.rst]: done

File: doc/source/tune-advance-tutorial.rst (hunk 12 above)
Comment (hhbyyh) [reply doc/source/tune-advance-tutorial.rst]: done
